zilliqa/src/eth_helpers.rs

```diff
@@ -0,0 +1,68 @@
+use ethabi::{ParamType, Token};
+
+pub(crate) fn extract_revert_msg(encoded: &[u8]) -> String {
+    // It could be either REVERT = keccka256("Error(string)") or PANIC = keccka256("Panic(uint256)")
+    // See https://docs.soliditylang.org/en/latest/control-structures.html#revert
+
+    // Try to decode revert message provided by EVM
+    const REVERT_SELECTOR: [u8; 4] = [0x08, 0xc3, 0x79, 0xa0];
+    const PANIC_SELECTOR: [u8; 4] = [0x4e, 0x48, 0x7b, 0x71];
+
+    let generic_error = String::from("execution reverted");
+    // At least 4 elemenets + anything to decode
+    if encoded.len() < REVERT_SELECTOR.len() + 1 {
+        return generic_error;
+    }
+    // It is safe since we check the size above
+    let prefix: [u8; 4] = unsafe { *encoded.as_ptr().cast::<[u8; 4]>() };
```

Comment:
Can we avoid the `unsafe` here? I doubt its necessary for performance.

You could do something like
```rust
let decoded_error = if encoded.starts_with(&REVERT_SELECTOR) {
  ethabi::decode(&[ParamType::String], &encoded[4..])
} else if encoded.starts_with(&PANIC_SELECTOR) {
  ethabi::decode(&[ParamType::Uint(256)], &encoded[4..])
} else {
  return generic_error;
};
let Ok(decoded_error) = decoded_error else { return generic_error; };
// etc.
```
?

Reply:
Also should we consider `ethabi::decode_whole` instead of `ethabi::decode`?